File: web/overviewplan/src/model/SiteplanObject.ts
/**
 * Copyright (c) 2022 DB Netz AG and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 */

import { checkInstance } from '@/util/ObjectExtension'

export interface ObjectColor {
    id: string
    color: number[]
}

export function defaultObjectColorObj () : ObjectColor {
  return {
    id: 'color',
    color: [0, 0, 0]
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isInstanceOfObjectColor (x: any): boolean {
  return checkInstance(x, defaultObjectColorObj())
}

export default interface SiteplanObject {
    guid: string
    // Not sent from Java API
    objectColors: ObjectColor[]
}

export function defaultSiteplanObj (): SiteplanObject {
  return {
    guid: 'abc',
    objectColors: [defaultObjectColorObj()]
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isInstanceOfSiteplanOject (object: any): boolean {
  return checkInstance(object, defaultSiteplanObj())
}

export function getLabelColor (object: SiteplanObject | ObjectColor[]): number[] | undefined {
  return getColor(object, 'label')
}

export function getColor (object: SiteplanObject | ObjectColor[], id?: string): number[] | undefined {
  if (object === undefined) {
    return undefined
  }

  let colorArr: ObjectColor[] = []
  if (isInstanceOfObjectColor(object)) {
    colorArr = (object as ObjectColor[])
  } else if (isInstanceOfSiteplanOject(object)) {
    colorArr = (object as SiteplanObject).objectColors
  } else {
    return undefined
  }

  if (!id) {
    return colorArr.find(ele => ele.id === 'feature')?.color
  }

  return colorArr.find(ele => ele.id === id)?.color
}

File: java/bundles/org.eclipse.set.model.siteplan/src/org/eclipse/set/model/siteplan/SiteplanObject.java
/**
 * Copyright (c) 2021 DB Netz AG and others.
 * 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 */
package org.eclipse.set.model.siteplan;

import org.eclipse.emf.ecore.EObject;

/**
 * <!-- begin-user-doc --> A representation of the model object
 * '<em><b>Object</b></em>'. <!-- end-user-doc -->
 *
 * <p>
 * The following features are supported:
 * </p>
 * <ul>
 * <li>{@link org.eclipse.set.model.siteplan.SiteplanObject#getGuid
 * <em>Guid</em>}</li>
 * </ul>
 *
 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getSiteplanObject()
 * @model
 * @generated
 */
public interface SiteplanObject extends EObject {
	/**
	 * Returns the value of the '<em><b>Guid</b></em>' attribute. <!--
	 * begin-user-doc --> <!-- end-user-doc -->
	 * 
	 * @return the value of the '<em>Guid</em>' attribute.
	 * @see #setGuid(String)
	 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getSiteplanObject_Guid()
	 * @model
	 * @generated
	 */
	String getGuid();

	/**
	 * Sets the value of the
	 * '{@link org.eclipse.set.model.siteplan.SiteplanObject#getGuid
	 * <em>Guid</em>}' attribute. <!-- begin-user-doc --> <!-- end-user-doc -->
	 * 
	 * @param value
	 *            the new value of the '<em>Guid</em>' attribute.
	 * @see #getGuid()
	 * @generated
	 */
	void setGuid(String value);

} // SiteplanObject
